### PR TITLE
Axelar hook

### PR DESCRIPTION
**Pull request by NicholasDotSol**
### Description

Added V1 Axelar post dispatch hook + tests

### CODE REVIEW

File: solidity/contracts/hooks/axelar/AxelarHook.sol [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        );
+
+        uint256 quote;
+        assembly {
+            // Copy the custom metadata to memory
+            // The '0x20' adds an offset for the length field in memory
+            calldatacopy(0x20, customMetadata.offset, customMetadata.length)
+            // Load the data from memory into 'number'
+            quote := mload(0x20)
+        }
```
Comment (yorhodes) [inline solidity/contracts/hooks/axelar/AxelarHook.sol:114]:
```suggestion
        uint256 quote = customMedatata[offset:offset+length];
```

File: solidity/contracts/hooks/axelar/AxelarHook.sol [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        return uint8(IPostDispatchHook.Types.AXELAR);
+    }
+
+    /**
+     * @notice Returns whether the hook supports metadata
+     * @return true the hook supports metadata
+     */
+    function supportsMetadata(bytes calldata) external pure returns (bool) {
+        return true;
+    }
```
Comment (yorhodes) [inline solidity/contracts/hooks/axelar/AxelarHook.sol:65]: could make this support both standard by "guessing" the axelar amount and your special variant with custom metadata

File: solidity/contracts/hooks/axelar/AxelarHook.sol [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+contract AxelarHook is IPostDispatchHook {
+    using StandardHookMetadata for bytes;
+    using Message for bytes;
+
+    IAxelarGasService public immutable AXELAR_GAS_SERVICE;
+    IAxelarGateway public immutable AXELAR_GATEWAY;
+    string public DESTINATION_CHAIN;
+    string public DESTINATION_CONTRACT;
+    bytes GMP_CALL_CODE;
```
Comment (yorhodes) [inline solidity/contracts/hooks/axelar/AxelarHook.sol:36]: interesting, would be curious to learn more about what this is

Comment (yorhodes) [reply solidity/contracts/hooks/axelar/AxelarHook.sol:36]: we also have some clever tricks for storing immutable bytes elsewhere in the codebase
see `MetaProxy`

File: solidity/contracts/hooks/wormhole/WormholeHook.sol [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+interface IWormhole {
+    function publishMessage(
+        uint32 nonce,
+        bytes memory payload,
+        uint8 consistencyLevel
+    ) external payable returns (uint64 sequence);
+}
+
+contract WormholeHook is IPostDispatchHook {
+    IWormhole public wormhole;
```
Comment (yorhodes) [inline solidity/contracts/hooks/wormhole/WormholeHook.sol:18]: immutable?

File: solidity/contracts/hooks/axelar/AxelarHook.sol (hunk 1 above)
Comment (yorhodes) [reply solidity/contracts/hooks/axelar/AxelarHook.sol:114]: https://soliditylang.org/blog/2020/05/26/array-slices/

File: solidity/contracts/hooks/wormhole/WormholeHook.sol (hunk 4 above)
Comment (NicholasDotSol) [reply solidity/contracts/hooks/wormhole/WormholeHook.sol:18]: Apologies the wormhole hook wasn't meant to be included here it's part of an already existing PR

File: solidity/contracts/hooks/axelar/AxelarHook.sol (hunk 3 above)
Comment (NicholasDotSol) [reply solidity/contracts/hooks/axelar/AxelarHook.sol:36]: this is just any data required by the GMP call to get the paylod to the receiver chain ISM. will change it's name

File: solidity/contracts/hooks/axelar/AxelarHook.sol (hunk 2 above)
Comment (NicholasDotSol) [reply solidity/contracts/hooks/axelar/AxelarHook.sol:65]: preferably we should receive an exact amount

File: solidity/contracts/hooks/axelar/AxelarHook.sol (hunk 1 above)
Comment (NicholasDotSol) [reply solidity/contracts/hooks/axelar/AxelarHook.sol:114]: this is good :)

Comment (NicholasDotSol) [reply solidity/contracts/hooks/axelar/AxelarHook.sol:114]: changing to slicing

File: solidity/contracts/hooks/axelar/AxelarHook.sol [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        uint256 quote = bridgeMetadata.axelarGasPayment();
+        require(quote > 0, "No Axelar Payment Received");
+
+        return quote;
+    }
+
+    function _formatPayload(
+        bytes calldata message
+    ) internal view returns (bytes memory) {
+        return abi.encodePacked(GMP_CALL_DATA, message.id());
```
Comment (NeverHappened) [inline solidity/contracts/hooks/axelar/AxelarHook.sol:118]: Okay, this should be something like this:
```abi.encodePacked(VERSION, payload)```
where VERSION is 4 bytes (like 0x0001?)
and payload is 
```
{
  "submit_meta": {
     "id": "hyperlane_message_id"
  }
}
```

Comment (NicholasDotSol) [reply solidity/contracts/hooks/axelar/AxelarHook.sol:118]: modifying the contract to compute the payload internally instead for clarity on testing. Good chance we revert back to pre-computed GMP_CALL_DATA generated on construct for some gas savings

Comment (NeverHappened) [reply solidity/contracts/hooks/axelar/AxelarHook.sol]: just for the record, [this comment](https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3050#discussion_r1428078595) is now obsolete, we agreed on using different version of payload encoding (version 1)

**Pull request closed without merging**